File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        
        def dfs(root, count):
            if root == None:
                return count

            return max(
                dfs(root.left, count + 1), 
                dfs(root.right, count + 1)
            )
        
        left_height = dfs(root.left, 0)
        right_height = dfs(root.right, 0)

        diameter = left_height + right_height
        
        return max(diameter, max(self.diameterOfBinaryTree(root.left), self.diameterOfBinaryTree(root.right)))

File: Data Structures & Algorithms/max-water-container/submission-1.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int i = 0;
        int j = heights.size() - 1;

        int maxWater = 0;
        while (i < j) {
            int left = heights[i];
            int right = heights[j];

            int area = (j - i) * std::min(left, right);
            maxWater = std::max(maxWater, area);

            if (left < right) {
                i++;
            } else {
                j--;
            }
        }

        return maxWater;
    }
};
